dotnet/src/webdriver/DevTools/DevToolsSession.cs

```diff
@@ -437,116 +413,38 @@ private void OnTargetDetached(object sender, TargetDetachedEventArgs e)
             }
         }
 
-        private async Task ReceiveMessage()
+        private async Task ShutdownMessageQueue()
         {
-            var cancellationToken = this.receiveCancellationToken.Token;
-            try
+            // Attempt to wait for the channel to empty before marking the
+            // writer as complete and waiting for the monitor task to end.
+            while (this.messageQueue.Reader.TryPeek(out _))
             {
-                var buffer = WebSocket.CreateClientBuffer(1024, 1024);
-                while (this.sessionSocket.State != WebSocketState.Closed && !cancellationToken.IsCancellationRequested)
-                {
-                    WebSocketReceiveResult result = await this.sessionSocket.ReceiveAsync(buffer, cancellationToken);
-                    if (!cancellationToken.IsCancellationRequested)
-                    {
-                        if (result.MessageType == WebSocketMessageType.Close && this.sessionSocket.State == WebSocketState.CloseReceived)
-                        {
-                            LogTrace("Got WebSocket close message from browser");
-                            await this.sessionSocket.CloseOutputAsync(WebSocketCloseStatus.NormalClosure, string.Empty, cancellationToken);
-                        }
-                    }
-
-                    if (this.sessionSocket.State == WebSocketState.Open && result.MessageType != WebSocketMessageType.Close)
-                    {
-                        using (var stream = new MemoryStream())
-                        {
-                            stream.Write(buffer.Array, 0, result.Count);
-                            while (!result.EndOfMessage)
-                            {
-                                result = await this.sessionSocket.ReceiveAsync(buffer, cancellationToken);
-                                stream.Write(buffer.Array, 0, result.Count);
-                            }
-
-                            stream.Seek(0, SeekOrigin.Begin);
-                            using (var reader = new StreamReader(stream, Encoding.UTF8))
-                            {
-                                string message = reader.ReadToEnd();
-
-                                // fire and forget
-                                // TODO: we need implement some kind of queue
-                                Task.Run(() => ProcessIncomingMessage(message));
-                            }
-                        }
-                    }
-                }
-            }
-            catch (OperationCanceledException)
-            {
-            }
-            catch (WebSocketException)
-            {
-            }
-            finally
-            {
-                this.sessionSocket.Dispose();
-                this.sessionSocket = null;
+                await Task.Delay(TimeSpan.FromMilliseconds(10));
             }
+
+            this.messageQueue.Writer.Complete();
+            this.messageQueueMonitorTask.Wait();
```

Comment:
`Wait` may introduce potential deadlock. Better `GetAwaiter().GetResult()` or even `await`.

Reply:
Given how and when this code is called, I believe the chances of deadlock are quite minimal, but sure, this doesn’t really fit the idiomatic pattern, so I’m happy to change it. Using `await` here is not the best choice because we want to retain the synchronous API (and yes, I’ve read all the literature about not mixing sync/async APIs;, that ship has sailed as the library API is largely synchronous by design).